src/distributions/uniform.rs

```diff
@@ -534,59 +533,110 @@ macro_rules! uniform_float_impl {
 
             fn new(low: Self::X, high: Self::X) -> Self {
                 assert!(low < high, "Uniform::new called with `low >= high`");
-                let scale = high - low;
-                let offset = low - scale;
-                UniformFloat {
-                    scale: scale,
-                    offset: offset,
+                assert!(low.is_finite() && high.is_finite(), "Uniform::new called with non-finite boundaries");
+
+                let mut scale = high - low;
+
+                let max_rand = (::core::$uty::MAX >> $bits_to_discard)
+                               .into_float_with_exponent(0) - 1.0;
+
+                while scale * max_rand + low >= high {
+                    scale = <$ty>::from_bits(scale.to_bits() - 1);
                 }
+
+                debug_assert!(scale >= 0.0);
+
+                UniformFloat { low, scale }
             }
 
             fn new_inclusive(low: Self::X, high: Self::X) -> Self {
-                assert!(low <= high,
-                        "Uniform::new_inclusive called with `low > high`");
-                let scale = high - low;
-                let offset = low - scale;
-                UniformFloat {
-                    scale: scale,
-                    offset: offset,
+                assert!(low <= high, "Uniform::new_inclusive called with `low > high`");
+                assert!(low.is_finite() && high.is_finite(), "Uniform::new_inclusive called with non-finite boundaries");
+
+                let mut scale = high - low;
+
+                let max_rand = (::core::$uty::MAX >> $bits_to_discard)
+                               .into_float_with_exponent(0) - 1.0;
+
+                if scale * max_rand + low > high {
+                    // Adjust scale lower until the max value we can generate is `high`
+                    while scale * max_rand + low > high {
+                        scale = <$ty>::from_bits(scale.to_bits() - 1);
+                    }
+                } else if high != low {
+                    // No need to loop here since even if `scale` is strictly not as big
+                    // as it could be, we'll still generate good values.
+                    // But give it one try to increase our chances that
+                    // Uniform::new_inclusive is actually inclusive of the high value.
+                    let new_scale = <$ty>::from_bits(scale.to_bits() + 1);
```

Comment:
I don't think there's sufficient justification for this code — it's extra complexity to *slightly* improve the range in a case where it already meets all *required* constraints.

Reply:
I don't have a strong opinion. But without this code it seems like there is very little practical difference between `new` and `new_inclusive`. I.e. it's very unlikely that the `high` value can actually be generated. It's arguably still very unlikely, but at least this code gives it a try.

But again, I'm not super opinionated on this. I'm not sure what the use case is for inclusive ranges for floats, but obviously the API requires it, so we need to stick something in this function.

Reply:
Are there many cases where this code will do something useful?

Reply:
I'm not sure how to measure usefulness given that I don't know of use cases for inclusive ranges for floats.

Given that there was a comment in the code which pointed out that `new` and `new_inclusive` generated the same values, do we have any use cases in mind where that is a problem?

Reply:
Another approach here could be to implement `new_inclusive` as

```rust
let max_rand = (::core::$uty::MAX >> $bits_to_discard)
               .into_float_with_exponent(0) - 1.0;

let mut scale = (high - low) / max_rand;

// Adjust scale lower until the max value we can generate is `high`
while scale * max_rand + low > high {
    scale = <$ty>::from_bits(scale.to_bits() - 1);
}

debug_assert!(scale >= 0.0);

UniformFloat { low, scale }
```

I don't know if the `/ max_rand` actually makes a difference. But from a mathematical point of view it at least means that we increase the chance of being able to generate `high`.

Reply:
I think `1.0 / max_rand > 1.0` so it probably does do something. Can the compiler optimise to multiplication or would it be better to write `(high - low) * (1.0 / max_rand)`?

The implementation is simpler at least — unless the current version has a clear advantage this looks preferable.

Reply:
Let's use this approach rather than the one I'm using right now.

Is the loss of precision writing `(high - low) * (1.0 / max_rand)` really worth making the ctor faster? I.e. is a ctor performance really a priority?

Reply:
I don't know. Possibly not, but I think `new_inclusive` may get used for single-usage constructors quite a bit. But anyway, we're not really losing precision since the `while` loop then reduces `scale`.